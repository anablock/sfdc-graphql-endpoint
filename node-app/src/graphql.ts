import {
    GraphQLBoolean,
    GraphQLEnumType,
    GraphQLFieldConfig,
    GraphQLFieldResolver,
    GraphQLFloat,
    GraphQLID,
    GraphQLInputFieldConfig,
    GraphQLInputObjectType,
    GraphQLInt,
    GraphQLList,
    GraphQLNonNull,
    GraphQLObjectType,
    GraphQLOutputType,
    GraphQLScalarType,
    GraphQLSchema,
    GraphQLString,
} from 'graphql';

import {
    Entity,
    Field,
    FieldType,
    isPolymorphicReference,
    isReferenceField,
    isScalarField,
    ReferenceField,
    ScalarField,
} from './entity.js';

declare module 'graphql' {
    interface GraphQLObjectTypeExtensions {
        sfdc?: Entity;
    }

    interface GraphQLFieldExtensions<_TSource, _TContext, _TArgs = any> {
        sfdc?: Field;
    }
}

interface SchemaConfig<Context> {
    entities: Entity[];
    resolvers?: {
        query?: (entity: Entity) => GraphQLFieldResolver<unknown, Context>;
        queryMany?: (entity: Entity) => GraphQLFieldResolver<unknown, Context>;
    };
}

interface State {
    config: SchemaConfig<any>;
    types: Map<string, GraphQLObjectType>;
    columnExprTypes: Map<string, GraphQLInputObjectType>;
    orderByTypes: Map<string, GraphQLList<GraphQLInputObjectType>>;
}

const SCALAR_TYPES_MAPPING: { [type in ScalarField['type']]: GraphQLScalarType } = {
    // Builtin GraphQL scalars
    [FieldType.ID]: GraphQLID,
    [FieldType.STRING]: GraphQLString,
    [FieldType.BOOLEAN]: GraphQLBoolean,
    [FieldType.INT]: GraphQLInt,
    [FieldType.FLOAT]: GraphQLFloat,

    // Custom SFDC types
    [FieldType.DATE]: new GraphQLScalarType({ name: 'Date' }),
    [FieldType.DATETIME]: new GraphQLScalarType({ name: 'DateTime' }),
    [FieldType.BASE64]: new GraphQLScalarType({ name: 'Base64' }),
    [FieldType.CURRENCY]: new GraphQLScalarType({ name: 'Currency' }),
    [FieldType.TEXTAREA]: new GraphQLScalarType({ name: 'TextArea' }),
    [FieldType.PERCENT]: new GraphQLScalarType({ name: 'Percent' }),
    [FieldType.PHONE]: new GraphQLScalarType({ name: 'Phone' }),
    [FieldType.URL]: new GraphQLScalarType({ name: 'URL' }),
    [FieldType.EMAIL]: new GraphQLScalarType({ name: 'Email' }),
    [FieldType.COMBOBOX]: new GraphQLScalarType({ name: 'Combobox' }),
    [FieldType.ANY_TYPE]: new GraphQLScalarType({ name: 'AnyType' }),
    [FieldType.PICKLIST]: new GraphQLScalarType({ name: 'Picklist' }),
    [FieldType.MULTI_PICKLIST]: new GraphQLScalarType({ name: 'MultiPicklist' }),
    [FieldType.ADDRESS]: new GraphQLScalarType({ name: 'Address' }),
    [FieldType.LOCATION]: new GraphQLScalarType({ name: 'Location' }),
};

const OPERATOR_INPUT_TYPES_MAPPING = Object.fromEntries(
    Object.entries(SCALAR_TYPES_MAPPING).map(([key, type]) => [key, createInputOperator({ type })]),
) as { [type in ScalarField['type']]: GraphQLInputObjectType };

const BY_ID_INPUT_ARGS = {
    id: {
        type: GraphQLID,
    },
};

const PAGINATION_INPUT_ARGS = {
    limit: {
        type: new GraphQLNonNull(GraphQLInt),
    },
    offset: {
        type: GraphQLInt,
    },
};

export type GraphQLSortOrderValue =
    | 'ASC'
    | 'DESC'
    | 'ASC_NULLS_FIRST'
    | 'ASC_NULLS_LAST'
    | 'DESC_NULLS_FIRST'
    | 'DESC_NULLS_LAST';

const ORDER_BY_ENUM_TYPE_VALUES: { [name in GraphQLSortOrderValue]: {} } = {
    ASC: {},
    DESC: {},
    ASC_NULLS_FIRST: {},
    ASC_NULLS_LAST: {},
    DESC_NULLS_FIRST: {},
    DESC_NULLS_LAST: {},
};

const ORDER_BY_ENUM_TYPE = new GraphQLEnumType({
    name: 'OrderByEnum',
    values: ORDER_BY_ENUM_TYPE_VALUES,
});

export function entitiesToSchema<Context>(config: SchemaConfig<Context>): GraphQLSchema {
    const state: State = {
        config,
        types: new Map(),
        orderByTypes: new Map(),
        columnExprTypes: new Map(),
    };

    for (const entity of config.entities) {
        const type = createGraphQLEntityType(state, entity);
        state.types.set(entity.sfdcName, type);
    }

    const query = createQuery(state, config.entities);

    return new GraphQLSchema({
        query,
        types: Array.from(state.types.values()),
    });
}

function createGraphQLEntityType(state: State, entity: Entity): GraphQLObjectType {
    const { gqlName, fields } = entity;

    return new GraphQLObjectType({
        name: gqlName,
        fields: () => {
            return Object.fromEntries(
                fields.map((field) => [field.gqlName, createGraphQLEntityField(state, field)]),
            );
        },
        extensions: {
            sfdc: entity,
        },
    });
}

function createGraphQLEntityField(state: State, field: Field): GraphQLFieldConfig<any, unknown> {
    let type: GraphQLOutputType;
    let resolve: GraphQLFieldResolver<any, unknown> = (source) => {
        return source[field.sfdcName];
    };

    if (isScalarField(field)) {
        type = SCALAR_TYPES_MAPPING[field.type];
    } else if (isReferenceField(field)) {
        // If the the referenced entity is part of the graph, make the field resolve to the
        // associated GraphQL type. Otherwise make the field resolve object id.
        if (state.types.has(field.sfdcReferencedEntityName)) {
            type = state.types.get(field.sfdcReferencedEntityName)!;
            resolve = (source) => {
                return source[field.sfdcReferencedEntityName];
            };
        } else {
            // TODO: Add support for lookup that aren't part of the graph.
            type = GraphQLID;
        }
    } else {
        // TODO: Add support for polymorphic field lookups.
        type = GraphQLID;
    }

    if (!field.config.nillable) {
        type = new GraphQLNonNull(type);
    }

    return {
        type,
        resolve,
        extensions: {
            sfdc: field,
        },
    };
}

function createQuery(state: State, entities: Entity[]): GraphQLObjectType {
    return new GraphQLObjectType({
        name: 'Query',
        fields: () => {
            return Object.assign(
                {},
                ...entities.map((entity) => createEntityQueries(state, entity)),
            );
        },
    });
}

function createEntityQueries(
    state: State,
    entity: Entity,
): Record<string, GraphQLFieldConfig<unknown, unknown>> {
    const { gqlName, sfdcName } = entity;
    const { resolvers } = state.config;

    const type = state.types.get(sfdcName)!;

    const orderByInputType = new GraphQLList(
        new GraphQLInputObjectType({
            name: `${sfdcName}OrderBy`,
            fields: Object.fromEntries(
                entity.fields
                    .filter((field) => field.config.sortable)
                    .map((field) => [
                        field.gqlName,
                        {
                            type: isReferenceField(field)
                                ? state.orderByTypes.get(field.sfdcReferencedEntityName) ??
                                  ORDER_BY_ENUM_TYPE
                                : ORDER_BY_ENUM_TYPE,
                        },
                    ]),
            ),
        }),
    );
    state.orderByTypes.set(sfdcName, orderByInputType);

    const columnExprInputType: GraphQLInputObjectType = new GraphQLInputObjectType({
        name: `${sfdcName}ColumnExpr`,
        fields: () => {
            return {
                _and: {
                    type: new GraphQLList(columnExprInputType),
                },
                _or: {
                    type: new GraphQLList(columnExprInputType),
                },
                ...Object.fromEntries(
                    entity.fields
                        .filter(
                            (field): field is ScalarField | ReferenceField =>
                                field.config.filterable && !isPolymorphicReference(field),
                        )
                        .map((field) => [
                            field.gqlName,
                            {
                                type: isReferenceField(field)
                                    ? state.columnExprTypes.get(field.sfdcReferencedEntityName) ??
                                      ORDER_BY_ENUM_TYPE
                                    : OPERATOR_INPUT_TYPES_MAPPING[field.type],
                            },
                        ]),
                ),
            };
        },
    });
    state.columnExprTypes.set(sfdcName, columnExprInputType);

    return {
        [gqlName]: {
            args: {
                ...PAGINATION_INPUT_ARGS,
                where: {
                    type: columnExprInputType,
                },
                order_by: {
                    type: orderByInputType,
                },
            },
            type: new GraphQLList(type),
            resolve: resolvers?.queryMany?.(entity),
        },
        [`${gqlName}_by_id`]: {
            args: BY_ID_INPUT_ARGS,
            type: type,
            resolve: resolvers?.query?.(entity),
        },
    };
}

function createInputOperator({ type }: { type: GraphQLScalarType }): GraphQLInputObjectType {
    const { name } = type;
    const fields: Record<string, GraphQLInputFieldConfig> = {
        _eq: { type },
        _neq: { type },
        _gt: { type },
        _lt: { type },
        _gte: { type },
        _lte: { type },
        _in: { type: new GraphQLList(type) },
        _nin: { type: new GraphQLList(type) },
    };

    if (name === 'MultiPicklist') {
        fields._includes = { type };
        fields._excludes = { type };
    } else if (name === 'Text') {
        fields._like = { type };
    }

    return new GraphQLInputObjectType({
        name: `${name}Operator`,
        fields,
    });
}
