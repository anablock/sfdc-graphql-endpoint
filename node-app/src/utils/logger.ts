export interface LoggerFn {
    (...args: unknown[]): void;
}

export interface Logger {
    info: LoggerFn;
    warn: LoggerFn;
    error: LoggerFn;
    debug: LoggerFn;
}
