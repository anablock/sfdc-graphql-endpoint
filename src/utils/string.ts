export function camelCase(str: string): string {
    return str[0].toLowerCase() + str.slice(1);
}
