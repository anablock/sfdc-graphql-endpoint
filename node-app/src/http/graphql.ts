import * as graphql from 'graphql';
import { GraphQLSchema, Source, DocumentNode } from 'graphql';
import { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';

import { SchemaBuilder } from '../schema-builder.js';
import { Graph } from '../graph.js';

import { LRU } from '../utils/lru.js';

import { Api } from '../sfdc/api.js';

import sfdcFastifyPlugin from './sfdc-plugin.js';

const ENTITIES = ['Account', 'User', 'Lead', 'Opportunity'];

interface GraphQLQueryString {
    query?: string;
    variables?: string;
    operationName?: string;
}

interface GraphQLParams {
    query?: string;
    variables?: Record<string, unknown>;
    operationName?: string;
}

export async function graphqlFastifyPlugin(fastify: FastifyInstance) {
    let schema: GraphQLSchema | undefined;
    const queryCache = new LRU<string, DocumentNode>(100);

    fastify.register(sfdcFastifyPlugin);

    fastify.get<{
        Params: GraphQLQueryString;
    }>('/graphql', async (request, response) => {
        if (request.headers.accept?.includes('text/html')) {
            return respondWithGraphIql(response);
        }

        let variables;
        try {
            variables = request.params.variables ? JSON.parse(request.params.variables) : {};
        } catch (error) {
            throw {
                statusCode: 400,
                message: 'Invalid GraphQL query.',
                errors: ['Failed to parse variables'],
            };
        }

        return executeQuery(request, {
            ...request.params,
            variables,
        });
    });

    fastify.post<{
        Body: GraphQLParams;
    }>('/graphql', async (request, response) => {
        return executeQuery(request, {
            ...request.body,
        });
    });

    async function executeQuery(
        request: FastifyRequest,
        params: GraphQLParams,
    ): Promise<graphql.ExecutionResult> {
        const { api, connection } = fastify.sfdc;
        const { query, variables, operationName } = params;

        if (!schema) {
            request.log.info('Fetching graphQL schema');
            schema = await buildSchema(api);
        }

        if (!query) {
            throw {
                statusCode: 400,
                message: 'Invalid GraphQL query.',
                errors: ['Missing query body'],
            };
        }

        const queryDocumentAst = parseAndValidateQuery(schema, query);

        return graphql.execute({
            schema,
            operationName,
            variableValues: variables,
            document: queryDocumentAst,
            contextValue: {
                api,
                connection,
                logger: request.log
            },
        });
    }

    function parseAndValidateQuery(schema: GraphQLSchema, query: string): DocumentNode {
        let documentAst = queryCache.get(query);

        if (documentAst === null) {
            try {
                documentAst = graphql.parse(new Source(query));
            } catch (err: unknown) {
                throw { statusCode: 400, message: 'GraphQL syntax error.', errors: [err] };
            }

            const validationErrors = graphql.validate(schema, documentAst);
            if (validationErrors.length > 0) {
                throw {
                    statusCode: 400,
                    message: 'Invalid GraphQL query.',
                    errors: validationErrors,
                };
            }

            queryCache.set(query, documentAst);
        }

        return documentAst;
    }
}

async function buildSchema(api: Api): Promise<GraphQLSchema> {
    const sObjects = await Promise.all(ENTITIES.map((entity) => api.describeSObject(entity)));

    const graph = new Graph(sObjects);
    const schemaBuilder = new SchemaBuilder(graph);

    const schema = schemaBuilder.buildSchema();
    graphql.assertValidSchema(schema);

    return schema;
}

function respondWithGraphIql(response: FastifyReply) {
    response.type('text/html').send(`
        <html>
            <head>
                <title>Simple GraphiQL Example</title>
                <link href="https://unpkg.com/graphiql/graphiql.min.css" rel="stylesheet" />
            </head>
            <body style="margin: 0;">
                <div id="graphiql" style="height: 100vh;"></div>

                <script
                    crossorigin
                    src="https://unpkg.com/react/umd/react.production.min.js"
                ></script>
                <script
                    crossorigin
                    src="https://unpkg.com/react-dom/umd/react-dom.production.min.js"
                ></script>
                <script
                    crossorigin
                    src="https://unpkg.com/graphiql/graphiql.min.js"
                ></script>

                <script>
                    const fetcher = GraphiQL.createFetcher({ url: '/graphql' });

                    ReactDOM.render(
                        React.createElement(GraphiQL, { fetcher: fetcher }),
                        document.getElementById('graphiql'),
                    );
                </script>
            </body>
        </html>
    `);
}
