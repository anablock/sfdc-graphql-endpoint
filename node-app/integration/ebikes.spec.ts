import { FastifyInstance } from 'fastify';
import { buildClientSchema, getIntrospectionQuery, printSchema } from 'graphql';

import { gql, getApp, executeQuery } from './utils.js';

let app: FastifyInstance;
beforeAll(() => {
    app = getApp({
        entities: ['Product__c', 'Product_Family__c', 'Order_Item__c', 'Order__c'],
    });
});

describe('eBikes entity', () => {
    describe('introspection', () => {
        test('introspect schema', async () => {
            const { data } = await executeQuery({
                app,
                query: getIntrospectionQuery({
                    schemaDescription: true,
                }),
            });

            const schema = buildClientSchema(data);
            expect(printSchema(schema)).toMatchSnapshot();
        });
    });

    describe('get product list', () => {
        test('retrieve most expensive products', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(order_by: { price__c: DESC }, limit: 4) {
                            name
                            price__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products order by gender and price', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(order_by: { gender__c: ASC, price__c: ASC }, limit: 10) {
                            name
                            price__c
                            gender__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve more products', async () => {
            const query = gql`
                query getProducts($offset: Int = 0) {
                    Product__c(order_by: { price__c: DESC }, limit: 2, offset: $offset) {
                        name
                        price__c
                    }
                }
            `;

            const { data: initialProducts } = await executeQuery({
                app,
                query,
                variables: {
                    offset: 0,
                },
            });
            expect(initialProducts).toMatchSnapshot();

            const { data: moreProducts } = await executeQuery({
                app,
                query,
                variables: {
                    offset: 2,
                },
            });
            expect(moreProducts).toMatchSnapshot();
        });

        test('retrieve the products with suspensions', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { suspension__c: { _eq: true } }
                            order_by: { name: ASC }
                            limit: 10
                        ) {
                            name
                            price__c
                            suspension__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products in price range', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { price__c: { _gt: 1300, _lt: 1500 } }
                            order_by: { name: ASC }
                            limit: 10
                        ) {
                            name
                            price__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products by exact name', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { name: { _eq: "Neomov - Basic" } }
                            order_by: { name: ASC }
                            limit: 10
                        ) {
                            name
                            price__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products by matching name', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { name: { _like: "Neomov%" } }
                            order_by: { name: ASC }
                            limit: 10
                        ) {
                            name
                            price__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve products male or female products only', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { gender__c: { _in: ["Male", "Female"] } }
                            order_by: { name: ASC }
                            limit: 10
                        ) {
                            name
                            price__c
                            gender__c
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });

        test('retrieve the products with the "Rolling Mountain" family name', async () => {
            const { data } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(
                            where: { product_Family__r: { name: { _eq: "Rolling Mountain" } } }
                            order_by: { name: ASC }
                            limit: 10
                        ) {
                            name
                            price__c
                            product_Family__r {
                                name
                            }
                        }
                    }
                `,
            });
            expect(data).toMatchSnapshot();
        });
    });

    describe('get single product', () => {
        test('retrieve a specific product by id', async () => {
            const {
                data: { Product__c: productList },
            } = await executeQuery({
                app,
                query: gql`
                    {
                        Product__c(limit: 1) {
                            id
                            name
                        }
                    }
                `,
            });
            const {
                data: { Product__c_by_id: product },
            } = await executeQuery({
                app,
                query: gql`
                    query getProductById($id: ID!) {
                        Product__c_by_id(id: $id) {
                            id
                            name
                        }
                    }
                `,
                variables: {
                    id: productList[0].id,
                },
            });

            expect(product).toEqual(productList[0]);
        });
    });
});
