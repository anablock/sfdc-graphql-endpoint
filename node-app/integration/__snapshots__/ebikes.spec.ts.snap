// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eBikes entity get product list retrieve more products 1`] = `
Object {
  "Product__c": Array [
    Object {
      "name": "Rolling Mountain - Female",
      "price__c": 2300,
    },
    Object {
      "name": "Rolling Mountain - Male",
      "price__c": 2300,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve more products 2`] = `
Object {
  "Product__c": Array [
    Object {
      "name": "Neomov - Suspension",
      "price__c": 1850,
    },
    Object {
      "name": "Neomov - Female frame",
      "price__c": 1450,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve most expensive products 1`] = `
Object {
  "Product__c": Array [
    Object {
      "name": "Rolling Mountain - Female",
      "price__c": 2300,
    },
    Object {
      "name": "Rolling Mountain - Male",
      "price__c": 2300,
    },
    Object {
      "name": "Neomov - Suspension",
      "price__c": 1850,
    },
    Object {
      "name": "Neomov - Male frame",
      "price__c": 1450,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products by exact name 1`] = `
Object {
  "Product__c": Array [
    Object {
      "name": "Neomov - Basic",
      "price__c": 1450,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products by matching name 1`] = `
Object {
  "Product__c": Array [
    Object {
      "name": "Neomov - Basic",
      "price__c": 1450,
    },
    Object {
      "name": "Neomov - Female frame",
      "price__c": 1450,
    },
    Object {
      "name": "Neomov - Male frame",
      "price__c": 1450,
    },
    Object {
      "name": "Neomov - Suspension",
      "price__c": 1850,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products in price range 1`] = `
Object {
  "Product__c": Array [
    Object {
      "name": "Neomov - Basic",
      "price__c": 1450,
    },
    Object {
      "name": "Neomov - Female frame",
      "price__c": 1450,
    },
    Object {
      "name": "Neomov - Male frame",
      "price__c": 1450,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products male or female products only 1`] = `
Object {
  "Product__c": Array [
    Object {
      "gender__c": "Female",
      "name": "Neomov - Female frame",
      "price__c": 1450,
    },
    Object {
      "gender__c": "Male",
      "name": "Neomov - Male frame",
      "price__c": 1450,
    },
    Object {
      "gender__c": "Female",
      "name": "Rolling Mountain - Female",
      "price__c": 2300,
    },
    Object {
      "gender__c": "Male",
      "name": "Rolling Mountain - Male",
      "price__c": 2300,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve products order by gender and price 1`] = `
Object {
  "Product__c": Array [
    Object {
      "gender__c": "Unisex",
      "name": "Wisper foldable - Upgraded",
      "price__c": 1230,
    },
    Object {
      "gender__c": "Unisex",
      "name": "Wisper foldable",
      "price__c": 1230,
    },
    Object {
      "gender__c": "Unisex",
      "name": "Neomov - Basic",
      "price__c": 1450,
    },
    Object {
      "gender__c": "Unisex",
      "name": "Neomov - Suspension",
      "price__c": 1850,
    },
    Object {
      "gender__c": "Male",
      "name": "Neomov - Male frame",
      "price__c": 1450,
    },
    Object {
      "gender__c": "Male",
      "name": "Rolling Mountain - Male",
      "price__c": 2300,
    },
    Object {
      "gender__c": "Female",
      "name": "Neomov - Female frame",
      "price__c": 1450,
    },
    Object {
      "gender__c": "Female",
      "name": "Rolling Mountain - Female",
      "price__c": 2300,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve the products with suspensions 1`] = `
Object {
  "Product__c": Array [
    Object {
      "name": "Neomov - Suspension",
      "price__c": 1850,
      "suspension__c": true,
    },
    Object {
      "name": "Rolling Mountain - Female",
      "price__c": 2300,
      "suspension__c": true,
    },
    Object {
      "name": "Rolling Mountain - Male",
      "price__c": 2300,
      "suspension__c": true,
    },
  ],
}
`;

exports[`eBikes entity get product list retrieve the products with the "Rolling Mountain" family name 1`] = `
Object {
  "Product__c": Array [
    Object {
      "name": "Rolling Mountain - Female",
      "price__c": 2300,
      "product_Family__r": Object {
        "name": "Rolling Mountain",
      },
    },
  ],
}
`;

exports[`eBikes entity introspection introspect schema 1`] = `
"type Product__c {
  id: ID!
  owner: ID!
  isDeleted: Boolean!
  name: String
  createdDate: DateTime!
  createdBy: ID!
  lastModifiedDate: DateTime!
  lastModifiedBy: ID!
  systemModstamp: DateTime!
  lastViewedDate: DateTime
  lastReferencedDate: DateTime
  autonomy__c: Float
  description__c: TextArea
  gender__c: Picklist
  picture_URL__c: URL!
  price__c: Currency
  product_Family__r: Product_Family__c
  suspension__c: Boolean!
}

scalar DateTime

scalar TextArea

scalar Picklist

scalar URL

scalar Currency

type Product_Family__c {
  id: ID!
  owner: ID!
  isDeleted: Boolean!
  name: String
  createdDate: DateTime!
  createdBy: ID!
  lastModifiedDate: DateTime!
  lastModifiedBy: ID!
  systemModstamp: DateTime!
  lastViewedDate: DateTime
  lastReferencedDate: DateTime
  category__c: Picklist
  description__c: TextArea
}

type Order_Item__c {
  id: ID!
  isDeleted: Boolean!
  name: String!
  createdDate: DateTime!
  createdBy: ID!
  lastModifiedDate: DateTime!
  lastModifiedBy: ID!
  systemModstamp: DateTime!
  lastViewedDate: DateTime
  lastReferencedDate: DateTime
  order__r: Order__c!
  product__r: Product__c!
}

type Order__c {
  id: ID!
  owner: ID!
  isDeleted: Boolean!
  name: String!
  createdDate: DateTime!
  createdBy: ID!
  lastModifiedDate: DateTime!
  lastModifiedBy: ID!
  systemModstamp: DateTime!
  lastViewedDate: DateTime
  lastReferencedDate: DateTime
}

type Query {
  Product__c(limit: Int!, offset: Int, where: Product__cColumnExpr, order_by: [Product__cOrderBy]): [Product__c]
  Product__c_by_id(id: ID): Product__c
  Product_Family__c(limit: Int!, offset: Int, where: Product_Family__cColumnExpr, order_by: [Product_Family__cOrderBy]): [Product_Family__c]
  Product_Family__c_by_id(id: ID): Product_Family__c
  Order_Item__c(limit: Int!, offset: Int, where: Order_Item__cColumnExpr, order_by: [Order_Item__cOrderBy]): [Order_Item__c]
  Order_Item__c_by_id(id: ID): Order_Item__c
  Order__c(limit: Int!, offset: Int, where: Order__cColumnExpr, order_by: [Order__cOrderBy]): [Order__c]
  Order__c_by_id(id: ID): Order__c
}

input Product__cColumnExpr {
  _and: [Product__cColumnExpr]
  _or: [Product__cColumnExpr]
  id: IDOperator
  isDeleted: BooleanOperator
  name: StringOperator
  createdDate: DateTimeOperator
  createdBy: OrderByEnum
  lastModifiedDate: DateTimeOperator
  lastModifiedBy: OrderByEnum
  systemModstamp: DateTimeOperator
  lastViewedDate: DateTimeOperator
  lastReferencedDate: DateTimeOperator
  autonomy__c: FloatOperator
  gender__c: PicklistOperator
  picture_URL__c: URLOperator
  price__c: CurrencyOperator
  product_Family__r: Product_Family__cColumnExpr
  suspension__c: BooleanOperator
}

input IDOperator {
  _eq: ID
  _neq: ID
  _gt: ID
  _lt: ID
  _gte: ID
  _lte: ID
  _in: [ID]
  _nin: [ID]
}

input BooleanOperator {
  _eq: Boolean
  _neq: Boolean
  _gt: Boolean
  _lt: Boolean
  _gte: Boolean
  _lte: Boolean
  _in: [Boolean]
  _nin: [Boolean]
}

input StringOperator {
  _eq: String
  _neq: String
  _gt: String
  _lt: String
  _gte: String
  _lte: String
  _in: [String]
  _nin: [String]
  _like: String
}

input DateTimeOperator {
  _eq: DateTime
  _neq: DateTime
  _gt: DateTime
  _lt: DateTime
  _gte: DateTime
  _lte: DateTime
  _in: [DateTime]
  _nin: [DateTime]
}

enum OrderByEnum {
  ASC
  DESC
  ASC_NULLS_FIRST
  ASC_NULLS_LAST
  DESC_NULLS_FIRST
  DESC_NULLS_LAST
}

input FloatOperator {
  _eq: Float
  _neq: Float
  _gt: Float
  _lt: Float
  _gte: Float
  _lte: Float
  _in: [Float]
  _nin: [Float]
}

input PicklistOperator {
  _eq: Picklist
  _neq: Picklist
  _gt: Picklist
  _lt: Picklist
  _gte: Picklist
  _lte: Picklist
  _in: [Picklist]
  _nin: [Picklist]
}

input URLOperator {
  _eq: URL
  _neq: URL
  _gt: URL
  _lt: URL
  _gte: URL
  _lte: URL
  _in: [URL]
  _nin: [URL]
}

input CurrencyOperator {
  _eq: Currency
  _neq: Currency
  _gt: Currency
  _lt: Currency
  _gte: Currency
  _lte: Currency
  _in: [Currency]
  _nin: [Currency]
}

input Product_Family__cColumnExpr {
  _and: [Product_Family__cColumnExpr]
  _or: [Product_Family__cColumnExpr]
  id: IDOperator
  isDeleted: BooleanOperator
  name: StringOperator
  createdDate: DateTimeOperator
  createdBy: OrderByEnum
  lastModifiedDate: DateTimeOperator
  lastModifiedBy: OrderByEnum
  systemModstamp: DateTimeOperator
  lastViewedDate: DateTimeOperator
  lastReferencedDate: DateTimeOperator
  category__c: PicklistOperator
}

input Product__cOrderBy {
  id: OrderByEnum
  owner: OrderByEnum
  isDeleted: OrderByEnum
  name: OrderByEnum
  createdDate: OrderByEnum
  createdBy: OrderByEnum
  lastModifiedDate: OrderByEnum
  lastModifiedBy: OrderByEnum
  systemModstamp: OrderByEnum
  lastViewedDate: OrderByEnum
  lastReferencedDate: OrderByEnum
  autonomy__c: OrderByEnum
  gender__c: OrderByEnum
  picture_URL__c: OrderByEnum
  price__c: OrderByEnum
  product_Family__r: OrderByEnum
  suspension__c: OrderByEnum
}

input Product_Family__cOrderBy {
  id: OrderByEnum
  owner: OrderByEnum
  isDeleted: OrderByEnum
  name: OrderByEnum
  createdDate: OrderByEnum
  createdBy: OrderByEnum
  lastModifiedDate: OrderByEnum
  lastModifiedBy: OrderByEnum
  systemModstamp: OrderByEnum
  lastViewedDate: OrderByEnum
  lastReferencedDate: OrderByEnum
  category__c: OrderByEnum
}

input Order_Item__cColumnExpr {
  _and: [Order_Item__cColumnExpr]
  _or: [Order_Item__cColumnExpr]
  id: IDOperator
  isDeleted: BooleanOperator
  name: StringOperator
  createdDate: DateTimeOperator
  createdBy: OrderByEnum
  lastModifiedDate: DateTimeOperator
  lastModifiedBy: OrderByEnum
  systemModstamp: DateTimeOperator
  lastViewedDate: DateTimeOperator
  lastReferencedDate: DateTimeOperator
  order__r: Order__cColumnExpr
  product__r: Product__cColumnExpr
}

input Order__cColumnExpr {
  _and: [Order__cColumnExpr]
  _or: [Order__cColumnExpr]
  id: IDOperator
  isDeleted: BooleanOperator
  name: StringOperator
  createdDate: DateTimeOperator
  createdBy: OrderByEnum
  lastModifiedDate: DateTimeOperator
  lastModifiedBy: OrderByEnum
  systemModstamp: DateTimeOperator
  lastViewedDate: DateTimeOperator
  lastReferencedDate: DateTimeOperator
}

input Order_Item__cOrderBy {
  id: OrderByEnum
  isDeleted: OrderByEnum
  name: OrderByEnum
  createdDate: OrderByEnum
  createdBy: OrderByEnum
  lastModifiedDate: OrderByEnum
  lastModifiedBy: OrderByEnum
  systemModstamp: OrderByEnum
  lastViewedDate: OrderByEnum
  lastReferencedDate: OrderByEnum
  order__r: OrderByEnum
  product__r: [Product__cOrderBy]
}

input Order__cOrderBy {
  id: OrderByEnum
  owner: OrderByEnum
  isDeleted: OrderByEnum
  name: OrderByEnum
  createdDate: OrderByEnum
  createdBy: OrderByEnum
  lastModifiedDate: OrderByEnum
  lastModifiedBy: OrderByEnum
  systemModstamp: OrderByEnum
  lastViewedDate: OrderByEnum
  lastReferencedDate: OrderByEnum
}"
`;
